# ATACseq-Analysis
This repository provides a workflow for analyzing ATAC-seq libraries prepared using the OMNI ATAC-seq protocol as described in Corces et al., 2017. 
Cells were treated with either DMSO (control) or three separate doses of the p300 inhibitor, A-485. ATAC-seq peaks were analyzed at Six1 binding sites (identified with a CUT&Tag-seq experiment with antibodies targetting Six1). 
This analysis was a part of a larger set of analysis for my master's thesis project: https://ruor.uottawa.ca/items/b8f8bfa4-0d34-45a6-9f13-75bb4ac52a04

This is analysis was completed in three steps.

Step 1: atac-seq-processing.sh (Performed on compute cluster)
Step 2: atac-feature-counting.sh (Performed on compute cluster, interactive session)
Step 3: atac-differential.analysis.R (Performed on local computer)

# Step 1: atac-seq-processing.sh

FASTQ Preprocessing:
- Concatenates reads from different lanes.
- Performs quality control using FastQC before and after trimming.
- Trims reads using fastp to remove adapters and low-quality sections.

Read Alignment:
- Maps reads to a reference genome using STAR.
- Post-Alignment Processing:
- Marks duplicates with Picard.
- Indexes BAM files using samtools.

Peak Calling:
- Converts BAM files to BED format.
- Performs peak calling using MACS2.
- Removes blacklisted regions from peak calls.

Replicate Aggregation:
- Uses MSPC to aggregate peak calls across replicates.

Downstream Analysis:
- Utilizes R for formatting consensus peak files and further analysis.

Requirements:
fastp - Version: 0.20.0
STAR - Version: 2.7.0a
Picard - Version: 2.18.9
samtools - Version: 1.9
FastQC - Version: 0.11.9
GCC - Version: 7.3.0
GSL - Version: 2.5
R - Version: 3.6.0
Python - Version: 3.7.0 (for MACS2)
Scipy-stack (for MACS2)
bedtools - Version: 2.29.2
MACS2 (no specific version mentioned)
MSPC (no specific version mentioned)
dotnet-core - Version: 3.0.0 (for MSPC)

# Step 2: atac-feature-counting.sh
This script converts the narrowPeak file generated by MACS2 in the previous script into a format for feature counting. It then uses featureCounts to count ATAC-seq reads at Six1 binding sites (identified using CUT&Tag-seq with antibodies targeting Six1).

Requirements
R: Version 3.6.0 
R Libraries:
ATACseqQC
GenomicAlignments
GenomicRanges
BSgenome.Mmusculus.UCSC.mm9
magrittr
dplyr
Biostrings
Rsubread
rtracklayer

Modules:
nixpkgs/16.09
gcc/7.3.0
gsl/2.5

# atac-differential.analysis.R 
This script carries out the differential analysis of ATAC-seq peaks at Six1 binding sites. The analysis includes loading and preprocessing ATAC-seq data, filtering out low-quality peaks, and applying normalization methods such as TMM (Trimmed Mean of M-values) and RUVr (Remove Unwanted Variation) to adjust for batch effects. The project also generates PCA plots to evaluate the effectiveness of different normalization methods and creates heatmaps and boxplots to visualize the expression data across different conditions.

Requirements:
R: 4.3.2
R Libraries:
GenomicAlignments
GenomicRanges
DESeq2
edgeR
BSgenome.Mmusculus.UCSC.mm9
magrittr
dplyr
ggplot2
tidyr
ggfortify
ggrepel
pheatmap
ggpubr
gridExtra
RUVSeq
rtracklayer
reshape2

# Computational Requirements:
Compute Cluster: For preprocessing and alignment steps.
Local Computer: For differential analysis step.
